include/deal.II/matrix_free/tools.h

```diff
@@ -39,40 +39,55 @@ namespace MatrixFreeTools
   categorize_by_boundary_ids(
     const Triangulation<dim> &tria,
     AdditionalData &          additional_data,
-    const unsigned int        level = numbers::invalid_unsigned_int)
+    const unsigned int        level = numbers::invalid_unsigned_int);
+
+
+
+  // implementations
+
+
+
+  template <int dim, typename AdditionalData>
+  void
+  categorize_by_boundary_ids(const Triangulation<dim> &tria,
+                             AdditionalData &          additional_data,
+                             const unsigned int        level)
   {
-    bool is_mg = (level != numbers::invalid_unsigned_int);
+    // ... determine if we are on an active or a multigrid level
+    const bool is_mg = (level != numbers::invalid_unsigned_int);
 
-    // ... create list for the category of each cell
+    // ... create empty list for the category of each cell
     if (is_mg)
       additional_data.cell_vectorization_category.resize(
         std::distance(tria.begin(level), tria.end(level)));
     else
       additional_data.cell_vectorization_category.resize(tria.n_active_cells());
 
     // ... set up scaling factor
-    std::vector<unsigned int> factors(dim * 2);
-
-    const auto bids = tria.get_boundary_ids();
+    std::vector<unsigned int> factors(GeometryInfo<dim>::faces_per_cell);
 
-    std::map<unsigned int, unsigned int> bid_map;
-    for (const auto id : tria.get_boundary_ids())
-      bid_map[id] = bid_map.size() + 1;
+    auto bids = tria.get_boundary_ids();
+    std::sort(bids.begin(), bids.end());
 
     {
       unsigned int n_bids = bids.size() + 1;
       int          offset = 1;
-      for (unsigned int i = 0; i < dim * 2; i++, offset = offset * n_bids)
+      for (unsigned int i = 0; i < GeometryInfo<dim>::faces_per_cell;
+           i++, offset = offset * n_bids)
         factors[i] = offset;
     }
 
-    auto to_category = [&](auto &cell) {
+    const auto to_category = [&](const auto &cell) {
       unsigned int c_num = 0;
-      for (unsigned int i = 0; i < dim * 2; i++)
+      for (unsigned int i = 0; i < GeometryInfo<dim>::faces_per_cell; i++)
         {
           auto &face = *cell->face(i);
-          if (face.at_boundary())
-            c_num += factors[i] * bid_map[face.boundary_id()];
+          if (face.at_boundary() && !cell->has_periodic_neighbor(i))
```

Comment:
In `MatrixFree` we regard periodic faces as inner faces. Missing this check led to too many categories and utterly destroyed the performance for small problem sizes in 6D in `hyper.deal`.

Reply:
The "utterly destroyed [...] performance for small problem sizes" is a general property of categorization and not limited to 6D: If you have few cells locally and you e.g. sit in a corner of the domain, it can be hard to fil all SIMD lanes with the same category. The problematic part with cell-based loops is that the decision applies an `and` across all faces of a cell, so the different restrictions end up accumulating over essentially all work that gets done. It would be interesting to describe the "categorization efficiency" as the ratio between the minimal number of cell batches, `(n_local_cells + n_simd_lanes - 1)/n_simd_lanes` and the actual number of cell batches.

In the longer run, I believe we will need to find entry points for the GPU way (or data-parallel) way of doing things, aka running both conditions of a branch and masking.

Reply:
> "categorization efficiency" as the ratio between the minimal number of cell batches, (n_local_cells + n_simd_lanes - 1)/n_simd_lanes and the actual number of cell batches

I was thinking about the same thing. Such a metric would have helped me to find the reason for the slow down quicker.

> In the longer run, I believe we will need to find entry points for the GPU way (or data-parallel) way of doing things, aka running both conditions of a branch and masking.

I agree. But image how step-67 + ECL would look like in this case...

Reply:
> But image how step-67 + ECL would look like in this case...

That calls for a library function hiding the complex details behind a nice user interface. :wink: